packages/typespec-vscode/src/emit/emit.ts

```diff
@@ -0,0 +1,395 @@
+// import { TypeSpecConfig } from "@typespec/compiler";
+import path, { dirname } from "path";
+import vscode, { Uri } from "vscode";
+import { Executable } from "vscode-languageclient/node.js";
+import { StartFileName } from "../const.js";
+import logger from "../log/logger.js";
+import { InstallationAction, npmDependencyType, NpmUtil } from "../npm-utils.js";
+import {
+  getMainTspFile,
+  resolveTypeSpecCli,
+  toError,
+  toOutput,
+  TraverseMainTspFileInWorkspace,
+} from "../typespec-utils.js";
+import { ExecOutput, isFile, spawnExecution } from "../utils.js";
+import { EmitQuickPickItem, TypeSpecProjectPickItem } from "./emit-quick-pick-item.js";
+import { Emitter, EmitterKind, getRegisterEmitters } from "./emitter.js";
+
+export async function doEmit(
+  context: vscode.ExtensionContext,
+  mainTspFile: string,
+  kind: EmitterKind,
+  overallProgress: vscode.Progress<{ message?: string; increment?: number }>,
+) {
+  if (!mainTspFile || !(await isFile(mainTspFile))) {
+    logger.info(
+      "Invalid typespec project. There is no main tsp file in the project. Generating Cancelled.",
+      [],
+      { showOutput: false, showPopup: true, progress: overallProgress },
+    );
+    return;
+  }
+
+  const baseDir = dirname(mainTspFile);
+
+  const toQuickPickItem = (e: Emitter): EmitQuickPickItem => {
+    return {
+      language: e.language,
+      package: e.package,
+      emitterKind: e.kind,
+      label: e.language,
+      detail: `Generate ${e.language} ${e.kind} code from ${e.package}`,
+      picked: false,
+      fromConfig: false,
+      iconPath: Uri.file(context.asAbsolutePath(`./icons/${e.language.toLowerCase()}.svg`)),
+    };
+  };
+
+  const registerEmitters = getRegisterEmitters(kind);
+  const all = [...registerEmitters].map((e) => toQuickPickItem(e));
+
+  const selectedEmitter = await vscode.window.showQuickPick<EmitQuickPickItem>(all, {
+    title: "Select a Language",
+    canPickMany: false,
+    placeHolder: "Pick a Language",
+    ignoreFocusOut: true,
+  });
+
+  if (!selectedEmitter) {
+    logger.info("No emitter selected. Generating Cancelled.", [], {
+      showOutput: false,
+      showPopup: false,
+      progress: overallProgress,
+    });
+    return;
+  }
+
+  /* TODO: verify the sdk runtime installation. */
+  /* inform to install needed runtime. */
+  const { valid, required } = await check(`${baseDir}/${StartFileName}`, selectedEmitter.package);
+  if (!valid) {
+    const toInstall = required.map((e) => e.name).join(", ");
+    await vscode.window
+      .showInformationMessage(
+        `Please install the required runtime for the selected emitters\n\n. ${toInstall}`,
+        "OK",
+      )
+      .then((selection) => {
+        if (selection === "OK") {
+          logger.info("Generating Cancelled.", [], {
+            showOutput: false,
+            showPopup: true,
```

Comment:
feels we don't need to popup again after user click "OK" in the first popup.

Reply:
actually i think the user experience here is still TBD and may change a lot, so i would suggest just remove this part to keep the code clean for now.

Reply:
removed.